File: src/Systems/Zig/singleton.zig
const std = @import("std");

const Registry = struct {
    count: u32 = 0,
};

var shared_registry = Registry{};

fn instance() *Registry {
    return &shared_registry;
}

pub fn main() void {
    const first = instance();
    const second = instance();
    first.count += 1;
    std.debug.print("same={s}\n", .{if (first == second) "true" else "false"});
    std.debug.print("count={d}\n", .{second.count});
}

File: src/Systems/Rust/singleton.rs
use std::sync::{Mutex, OnceLock};

static REGISTRY: OnceLock<Mutex<u32>> = OnceLock::new();

fn registry() -> &'static Mutex<u32> {
    REGISTRY.get_or_init(|| Mutex::new(0))
}

fn main() {
    let first = registry();
    let second = registry();
    *first.lock().expect("registry lock") += 1;
    println!("same={}", std::ptr::eq(first, second));
    println!("count={}", *second.lock().expect("registry lock"));
}
